openshift/console
# Update Web Terminal custom resources to use v1alpha2

### Description

This PR adds support for `v1alpha2` Web Terminal custom resources (DevWorkspaces) via a feature flag. If `v1alpha2` DevWorkspaces are available on the cluster, the console frontend will create those instead of current `v1alpha1` Web Terminals to continue to support web terminal functionality. 

Starting with OpenShift 4.8, we intend to make the Web Terminal operator depend on the DevWorkspace Operator via OLM. This change requires changes to the Web Terminal custom resource that gets created. Currently (i.e. OpenShift 4.7) the Web Terminal operator is deploying an old version of the DevWorkspace Operator under the hood.

### Related PRs
- Reconfigure WTO as depending on the DevWorkspace Operator: https://github.com/redhat-developer/web-terminal-operator/pull/80
- Begin building standalone operator for DevWorkspaces: https://github.com/devfile/devworkspace-operator/pull/438

### Testing this PR
**Note**: if the cluster being used for testing is OpenShift 4.7, it's necessary to include changes from https://github.com/openshift/console-operator/pull/513, which changed the backend console RBAC. The easiest way to do this is to run 
```
watch -n 1 kubectl apply -f manifests/03-rbac-role-cluster.yaml
```
from the console-operator repo while testing.

#### Testing that current functionality still works:
1. Install the Web Terminal Operator (version 1.2.1) via the OperatorHub
    - The 1.2.1 Web Terminal Operator uses the older v1alpha1 spec.
2. Check that terminals work for both cluster-admin and regular users.

#### Testing that new functionality works as expected:
1. There's no upgrade path from WTO 1.2.1 to 1.3.0, as the switch from a complete operator to one that depends on another operator is incompatible. As a result, existing instances of WTO 1.2.1 must be fully uninstalled:

    ```bash
    # 1. Ensure that all DevWorkspace Custom Resources are removed to avoid issues with finalizers
    # make sure depending objects are clean up as well
    kubectl delete devworkspaces.workspace.devfile.io --all-namespaces --all --wait
    kubectl delete workspaceroutings.controller.devfile.io --all-namespaces --all --wait
    kubectl delete components.controller.devfile.io --all-namespaces --all --wait
    # 2. Uninstall the Operator
    kubectl delete subscriptions.operators.coreos.com web-terminal -n openshift-operators --ignore-not-found
    WTO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("web-terminal.v1"))][0].metadata.name')
    kubectl delete csv ${WTO_CSV} -n openshift-operators
    # 3. Remove CRDs
    kubectl delete customresourcedefinitions.apiextensions.k8s.io workspaceroutings.controller.devfile.io
    kubectl delete customresourcedefinitions.apiextensions.k8s.io components.controller.devfile.io
    kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaces.workspace.devfile.io
    # 4. Remove DevWorkspace Webhook Server Deployment itself
    kubectl delete deployment/devworkspace-webhook-server -n openshift-operators
    # 5. Remove lingering service, secrets, and configmaps
    kubectl delete all --selector app.kubernetes.io/part-of=devworkspace-operator,app.kubernetes.io/name=devworkspace-webhook-server
    kubectl delete serviceaccounts devworkspace-webhook-server -n openshift-operators
    kubectl delete configmap devworkspace-controller -n openshift-operators
    kubectl delete clusterrole devworkspace-webhook-server
    kubectl delete clusterrolebinding devworkspace-webhook-server
    # 6. Remove mutating/validating webhooks configuration
    kubectl delete mutatingwebhookconfigurations controller.devfile.io
    kubectl delete validatingwebhookconfigurations controller.devfile.io
    ```
2. Apply custom CatalogSource to provide new version of Web Terminal Operator and dependency DevWorkspace Operator:

    ```yaml
    cat <<EOF | oc apply -f -
    apiVersion: operators.coreos.com/v1alpha1
    kind: CatalogSource
    metadata:
      name: custom-web-terminal-catalog
      namespace: openshift-marketplace
    spec:
      sourceType: grpc
      image: quay.io/amisevsk/devworkspace-demo-index:openshift-console
      publisher: Red Hat
      displayName: Web Terminal Operator Catalog
    EOF
    ```
3. Install Web Terminal Operator 1.3.0 from OperatorHub (it will be a separate entry, not in the default catalog)
4. Check that 1. Web Terminals work as expected for cluster-admin and regular users, and 2. that the CR created for Web Terminals has the `v1alpha2` version (`oc get dw -o yaml` from a web terminal)

### Cleanup
<details><summary>Cleaning up WTO 1.3.0 after testing</summary>
<p>

```bash
# -- 1 -- Uninstall WTO subscription and related CSV
kubectl delete subscriptions.operators.coreos.com web-terminal -n openshift-operators --ignore-not-found
WTO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("web-terminal.v1"))][0].metadata.name')
kubectl delete csv ${WTO_CSV} -n openshift-operators

# -- 2 -- Uinstall DWO subscription and related CSV
kubectl delete subscriptions.operators.coreos.com devworkspace-operator -n openshift-operators --ignore-not-found
DWO_CSV=$(kubectl get csv -o=json | jq -r '[.items[] | select (.metadata.name | contains("devworkspace-operator.v0"))][0].metadata.name')
kubectl delete csv ${DWO_CSV} -n openshift-operators

# -- 3 -- Make sure all related DWO objects are removed
  # 1. Ensure that all DevWorkspace Custom Resources are removed to avoid issues with finalizers
  # make sure depending objects are clean up as well
  kubectl delete devworkspaces.workspace.devfile.io --all-namespaces --all --wait
  kubectl delete devworkspaceroutings.controller.devfile.io --all-namespaces --all --wait
        kubectl delete devworkspacetemplates.workspace.devfile.io --all-namespaces --all --wait
  # 3. Remove CRDs
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaceroutings.controller.devfile.io
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspacetemplates.controller.devfile.io
  kubectl delete customresourcedefinitions.apiextensions.k8s.io devworkspaces.workspace.devfile.io
  # 4. Remove DevWorkspace Webhook Server Deployment itself
  kubectl delete deployment/devworkspace-webhook-server -n openshift-operators
  # 5. Remove lingering service, secrets, and configmaps
  kubectl delete all --selector app.kubernetes.io/part-of=devworkspace-operator,app.kubernetes.io/name=devworkspace-webhook-server
  kubectl delete serviceaccounts devworkspace-webhook-server -n openshift-operators
  kubectl delete configmap devworkspace-controller -n openshift-operators
  kubectl delete clusterrole devworkspace-webhook-server
  kubectl delete clusterrolebinding devworkspace-webhook-server
  # 6. Remove mutating/validating webhooks configuration
  kubectl delete mutatingwebhookconfigurations controller.devfile.io
  kubectl delete validatingwebhookconfigurations controller.devfile.io
```

</p>
</details>
